pages/workflow/WorkflowHandler.inc.php

```diff
@@ -239,23 +245,21 @@ protected function getStageDecisionTypes(int $stageId): array
                 }
                 break;
             case WORKFLOW_STAGE_ID_EDITING:
-                /** @var ReviewRoundDAO $reviewRoundDao */
-                $reviewRoundDao = DAORegistry::getDAO('ReviewRoundDAO');
-                $hasReviewRound = $reviewRoundDao->submissionHasReviewRound($submission->getId(), WORKFLOW_STAGE_ID_EXTERNAL_REVIEW);
                 $decisionTypes = [
                     new SendToProduction(),
-                    $hasReviewRound
-                        ? new BackToReview()
-                        : new BackToSubmissionFromCopyediting()
                 ];
                 break;
             case WORKFLOW_STAGE_ID_PRODUCTION:
-                $decisionTypes = [
-                    new BackToCopyediting(),
-                ];
+                $decisionTypes = [];
                 break;
         }
 
+        $decisionTypes = array_merge(
+            $decisionTypes,
+            $decisionRepository->getApplicableRetractableDecisionTypes($stageId, $submission, $reviewRoundId),
+            $decisionRepository->getApplicableRemovableDecisionTypes($stageId, $submission, $reviewRoundId)
+        );
```

Comment:
> With this code, each decision type goes to the database separately to determine if it canRetract() or canRemove(). That's a few extra db hits that we want to try to avoid.

@NateWr for those backable decision types that require a `$reviewRoundId`, if none passed it return `false` immediately. for example in the `BackToPreviousInternalReviewRound::canRetract`, we have 

```php
if (!$reviewRoundId) {
    return false;
}
```

for those it will do not require, it has no impact . Also the `$reviewRoundId`, we are determining it at the beginning as such 

```php
$reviewRoundId = (int) $request->getUserVar('reviewRoundId');
```

so I am bit confused about any extra DB hits ? 

> Having reviewed further, I do think that we'll want to move the logic out of the repository (eg - getApplicableXXXXXXDecisionTypes()) and into the workflow handler. The handler is the right place to put UI concerns like which buttons should be visible under which conditions.

I understand your point but that is not only the case here as we are determining which of the `backable/retractable` decision can be available for the editor to take action based on the current stage of submission . I think that kind of login can be in repository class . But in our case , it's the combination of both UI and applicable logic .

Reply:
> so I am bit confused about any extra DB hits ?

Here is an example of an extra DB hit in a `canRetract()` method:

https://github.com/pkp/pkp-lib/pull/8031/files#diff-d013eb089a0c96df5cf2a5c5901dbf1ce4b0ca58a49d0161ec22ba056eb77d0eR152

(In case the link doesn't work, it's `BackToExternalReview::152`).

> I understand your point but that is not only the case here as we are determining which of the backable/retractable decision can be available for the editor to take action based on the current stage of submission... But in our case , it's the combination of both UI and applicable logic .

My feeling is that application logic belongs in the `DecisionType::validate()` method. It looks to me like you're already calling `canRetract()` there, so that should cover the application logic...

Reply:
> I'm sorry, I might be [bikeshedding](https://en.wiktionary.org/wiki/bikeshedding) here.

@NateWr not at all . I understand your point to make it as efficient as possible as we want to avoid any possible future overhead as much as possible . 

> If you pass the full `ReviewRound` to the `canRetract()` method, you shouldn't need to do the DB lookup. Consider `BackToExternalReview`

I think I am missing something here . The back to `BackToExternalReview` which I have renamed as `BackToExternalReviewFromCopyediting` dont get any reviewround details when a submission is in the `copy editing stage` . I have checked and tested it . So in particular case , we dont have any `reviewRound` and look for . Also the implementation of the `canRetract()` method of as such 

```php
/**
 * Determine if can back out to external review stage from the copy editing stage
 */
public function canRetract(Submission $submission, ?int $reviewRoundId): bool
{
    /** @var ReviewRoundDAO $reviewRoundDao */
    $reviewRoundDao = DAORegistry::getDAO('ReviewRoundDAO');

    // if there is no external review round associated with it
    // can not back out to external review stage
    if (!$reviewRoundDao->submissionHasReviewRound($submission->getId(), WORKFLOW_STAGE_ID_EXTERNAL_REVIEW)) {
        return false;
    }

    return true;
}
```

as you can see , here we are also not trying to determined the state/status of any specific `reviewRound` either . we are trying to determined if it the submission has any review round at external review stage or not . 

Also if we check the decision `BackToPreviousExternalReviewRound` and the implementation of `canRetract` there , 
```php
/**
 * Determine if can back out form current external review round to previous external review round
 */
public function canRetract(Submission $submission, ?int $reviewRoundId): bool
{
    if (!$reviewRoundId) {
        return false;
    }

    /** @var ReviewRoundDAO $reviewRoundDao */
    $reviewRoundDao = DAORegistry::getDAO('ReviewRoundDAO');

    // if there is any submitted review by reviewer that is not cancelled
    // can not back out
    $confirmedReviewerIds = $this->getReviewerIds($submission->getId(), $reviewRoundId, self::REVIEW_ASSIGNMENT_CONFIRMED);
    if (count($confirmedReviewerIds) > 0) {
        return false;
    }

    // if this is the only round availabel
    // can not back out to previous round as there is none
    if ($reviewRoundDao->getReviewRoundCountBySubmissionId($submission->getId(), WORKFLOW_STAGE_ID_EXTERNAL_REVIEW) === 1) {
        return false;
    }

    // if there is any completed review by reviewer
    // can not back out
    $completedReviewAssignments = $this->getReviewAssignments($submission->getId(), $reviewRoundId, self::REVIEW_ASSIGNMENT_COMPLETED);
    if (count($completedReviewAssignments) > 0) {
        return false;
    }

    return true;
}
```

we can see that none of the check is has anything to do with the current reviweRound directly but determined assigned/completed/active review assignments etc . So not sure if passing the `reviewRound` DB object any impact for that . 

Or may be I am missing something here . I am very sorry if I am making things over complicated here.

> Maybe the decision should be `BackFromCopyediting` . 

Thanks a good idea and I thought about that when I first started working on this issue . But then I noticed that we have the implemented patterns for decisions here as `BackToXXXXXXFromXXXXX` so I went that way . Now if we make the changes only for the case of `back from the copy editing stage` , then we will introduce 2 different patterns which I think will make things bit complicated for us . So my suggestion is that we make all as `BackFromXXXXXX` or keep it as it is now . in future if we have more stage, then may be we can consider to refactor all of then . what you say ?

Reply:
> So my suggestion is that we make all as BackFromXXXXXX or keep it as it is now

Let's make changes to all of the decisions, even the ones you haven't touched. I think we'll simplify things a lot this way. 

I'll wait for that to be done and then revisit what the comments about `canRetract()`. I suspect once we've changed everything to `BackFromXXXXX`, that part will look a little bit different and we can re-evaluate then.

Reply:
@NateWr it's done . Please review it . If we make a final decision about it, then I will remove all the dead codes like previous decision pattern classes, translation keys, email templates etc .